migrate from msrestazure library to azure-identity

jwxd opened the pull request:
Addressing issue in  https://github.com/Hexadite/acs-keyvault-agent/issues/46
Updating the code to migrate to use azure-identity
Azure-identity sets a default timeout to 2sec when making http request to IMDS.

Will follow the test plan to make sure all authentication scenario are working properly


## Tests plan:
#### USE_MSI:
- [x] MSI Client id
- [x] MSI object id
- [x] MSI resource id
- [x] MSI with azure identity binding
#### USE_ENV:
- [x] service principal with env var
- [x] service principla using file
- [x] MSI with user_assigned_identity
- [x] client secret auth

All of the authentication mechanism has been tested working with new changes

Commit:
migrate to azure-identity

Commit:
fix user_assigned_identity method

Commit:
remove unused dependency

idanlevin commented on app/main.py:
```diff
@@ -72,35 +73,50 @@ def _parse_sp_file(self):
+            else:
+                # azure-identity migration
+                _logger.info('Assign azure-identity default env variable')
+                os.environ["AZURE_TENANT_ID"] = self.tenant_id
```
why do you need to change the os env?

jwxd replied:
https://azuresdkdocs.blob.core.windows.net/$web/python/azure-identity/latest/azure.identity.html#azure.identity.EnvironmentCredential
When using environment credential, azure identity takes it directly from the os env. 
In order to not break any of the existing workflow. I mapped the current os env into the one Azure identity is using

idanlevin replied:
the current place is in the `_parse_sp_file` which is only executed if the user did not specify "USE_ENV" (line #121), so we should not be dealing with OS env variables here, only in the `_parse_sp_env` function.

jwxd replied:
pushed a commit to remove os env var usage in case of parsing files.

idanlevin commented on app/main.py:
```diff
@@ -72,35 +73,50 @@ def _parse_sp_file(self):
 
-    def _get_client(self):
+        _logger.info('Assign new azure identity env variable')
+        os.environ["AZURE_TENANT_ID"] = os.environ["TENANT_ID"]
```
same question here

jwxd replied:
See answer above

idanlevin replied:
got it, but why read it twice from the env variable? you already have it in `self.tenant_id`, etc.

jwxd replied:
Change to use local variable in new commit

idanlevin commented on app/main.py:
```diff
@@ -72,35 +73,50 @@ def _parse_sp_file(self):
                 msi_client_id = os.environ["MSI_CLIENT_ID"]
                 _logger.info('Using client_id: %s', msi_client_id)
-                credentials = MSIAuthentication(resource=VAULT_RESOURCE_NAME, client_id=msi_client_id)
+                credentials = ManagedIdentityCredential(scopes=VAULT_RESOURCE_NAME, client_id=msi_client_id)
```
Can you pls point to the documentation of the change from "resource" to "scoped" parameter? couldn't find the documentation for it.

jwxd replied:
I remember reading one documentation mentioning scope replace resource but I couldn't find it anymore.
But these two line of code should prove that `scopes` and `resource` referring to the same thing
https://github.com/Azure/azure-sdk-for-python/blob/c8291ac6cb0dbd865da03a88dd2bcb9279e2c4a6/sdk/identity/azure-identity/tests/managed-identity-live/test_cloud_shell_async.py#L16
https://github.com/Azure/azure-sdk-for-python/blob/c8291ac6cb0dbd865da03a88dd2bcb9279e2c4a6/sdk/identity/azure-identity/azure/identity/_credentials/managed_identity.py#L75

idanlevin replied:
have you tried putting some wrong values to see if it does anything? I just want to make sure we don't break anything since it's undocumented

jwxd replied:
I think you are right. I tested with a incorrect VAULT_RESOURCE_NAME value but it still works.
I think the `scopes` parameters are not being used by the credential class at all. But only used in get_token() method.
I will need to revisit the the docs to see how to correctly use VAULT_RESOURCE_NAME in azure-identity.

idanlevin replied:
OK, please let me know when you do so and I'll merge, I don't want to merge until we realize if this is the correct usage.

jwxd replied:
It looks like [new SecretClient implementation](https://github.com/Azure/azure-sdk-for-python/blob/1dea0a408beb53b5733919e37f874b3600037fcd/sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/_shared/challenge_auth_policy.py#L75) that uses azure-identity creates default scope by parsing `VAULT_RESOURCE_NAME ` from the request url.
I think VAULT_RESOURCE_NAME is not be necessary in most cases when using azure-identity. 
Do you know if there are use cases when the VAULT_RESOURCE_NAME is not part of the vault url?

idanlevin replied:
@jwxd I think the VALUE_RESOURCE_NAME was different in some special clouds (AG, etc.)

jwxd replied:
I see, It seems like there might be some issue with azure-identity implementation that it's currently not supporting explicitly specify scopes. I have started a conversation with the team who developed the sdk. Link to the thread below.

jwxd replied:
btw I just checked our current use case in special clouds including AG. VAULT_RESOURCE_NAME is the postfix of the vault url as well. Can you verify if there are use cases where it is not?
If not I think it's safe to drop this env var. Otherwise, we will have to wait until Azure-identity support custom scopes

jwxd replied:
Hi Idan, we have tested the new azure-identity in our production environment without using VAULT_RESOURCE_NAME, and it works. Thus I removed the VAULT_RESOURCE_NAME usage from keyvault-agent. Please review

idanlevin commented on requirements.txt:
```diff
@@ -3,3 +3,6 @@ msrestazure==0.4.34
 adal==1.2.4
```
we should remove unused libraries (azure-kevault, msrestazure and adal) and make sure the tests work without them.

idanlevin commented:
hi @Michael-Sinz, @marloncalvo, @winterTTr, @abhaybhu10 can you please review and test your scenarios? this is a major change

idanlevin commented:
@jwxd thanks for contributing! had a few comments and want more folks to verify as this project doesn't have proper CI

winterTTr commented:
I will try to verify the MI case, since this is one of the major scenario my team is using, I will send the result later.

winterTTr commented on app/main.py:
```diff
@@ -112,18 +128,12 @@ def _get_client(self):
                 _logger.info('Using MSI')
                 # refer _parse_sp_file, potentially we could have mi client id from sp
                 if self.user_assigned_identity_id != "":
-                    _logger.info('Using client_id: %s', self.user_assigned_identity_id)
```
Please add this line back, since this means to support the case when there is multi MI binding on the same VM, so we need to explicitly config MI id to acs keyvault agent.  This is not mandatory, since if there is only single MI, we don't have to specify any id.  So this line is easier to keep the original log behavior the same.

jwxd replied:
pushed a commit to add this line back

winterTTr commented:
Test case passed from my side, my case can cover below scenario:

USE_MSI:
- [X] MSI Client id

USE_ENV:
- [X] service principla using file
- [X] MSI with user_assigned_identity

PS:
I add one comment, just add one of the log back since that's means to tell which client id we need to use.

Commit:
add logging for user_managed_identity

jwxd commented:
Can we expedite the reviewing and get this merged ASAP?

Commit:
Merged PR 7012042: migrate from msrestazure library to azure-identity

This is a fork of open source project acs-keyvault-agent.
Addressing issue in https://github.com/Hexadite/acs-keyvault-agent/issues/46

Updating the code to migrate to use azure-identity
Azure-identity sets a default timeout to 2sec when making http request to IMDS.

The changes has been tested and verify working

Related work items: #38164308

Commit:
remove deprecated dependency

Commit:
removed dependency

idanlevin commented:
hi @jwxd, the reason for it not being approved is that you haven't replied to all the PR comments yet.

jwxd commented:
@idanlevin  Forgot to resolve the conversation. I have addressed all comments.

> @jwxd thanks for contributing! had a few comments and want more folks to verify as this project doesn't have proper CI

Did you added any comments? I don't see any.

idanlevin commented:
> @idanlevin Forgot to resolve the conversation. I have addressed all comments.
> 
> > @jwxd thanks for contributing! had a few comments and want more folks to verify as this project doesn't have proper CI
> 
> Did you added any comments? I don't see any.

@jwxd sorry, I forgot to submit my comments, you should be able to see them now.

Commit:
fix usage of env var when using files

Commit:
Merge branch 'remove_deprecated_dependency' into azure_identity_migration

Commit:
remove incorrect usage of scopes parameter

amirschw commented on app/main.py:
```diff
@@ -72,35 +72,44 @@ def _parse_sp_file(self):
+                credentials = ManagedIdentityCredential(identity_config=identity)
             else:
-                credentials = MSIAuthentication(resource=VAULT_RESOURCE_NAME)
+                credentials = DefaultAzureCredential()
```
Is there a reason we don't use `ManagedIdentityCredential()` here? [DefaultAzureCredential](https://docs.microsoft.com/en-us/python/api/azure-identity/azure.identity.defaultazurecredential?view=azure-python) first tries to use a service principal configured by environment variables, and only after finding this option unavailable, tries to use a managed identity. So using `ManagedIdentityCredential`, if possible, should probably be slightly more performant.

jwxd replied:
Good point, I think using ManagedIdentityCredential make more sense here.

amirschw commented on app/main.py:
```diff
@@ -297,14 +308,16 @@ def grab_secrets(self):
             _logger.info('Retrieving all secrets from Key Vault.')
 
-            all_secrets = list(client.get_secrets(vault_base_url))
+            all_secrets = list(secret_client.list_properties_of_secrets())
```
Just making sure, since we're using some different methods to get / list secrets - these updates won't require any changes to the access policies needed by the keyvault-agent, right?

jwxd replied:
No changes to access policy.

amirschw commented on app/main.py:
```diff
@@ -72,35 +72,44 @@ def _parse_sp_file(self):
             if self.client_id == "msi" and self.client_secret == "msi":
                 self.user_assigned_identity_id = sp_data.get("userAssignedIdentityID", "")
-
+            
```
nit: trailing spaces
```suggestion

```

Commit:
change DefaultCredential to ManagedIdentity

Commit:
cleanup logging

Commit:
remove VAULT_RESOURCE_NAME env var usage
